### PR TITLE
Allow saving all attachments of a thread

### PR DESCRIPTION
**Pull request by fehnomenal**
Creates a new button in the `MediaOverviewActivity` to save all images of the conversation.
The diff is big because of different strings for single/multiple images to save.

### CODE REVIEW

**Comment by mcginty** [comment]
Cool with this change, a nice convenience button. It wouldn't hurt to have a "save all" icon (maybe just two stacked floppies) to make it clear this is a batch action, though.

**Comment by mcginty** [comment]
quick photoshop

ic_save_white_48dp

transparent version: https://cloud.githubusercontent.com/assets/373823/11358584/bb3a7b4c-9228-11e5-84c4-d7e1ba7539aa.png

honestly it kind of makes it look like an old timey large-format film camera

**Comment by Omikron43** [comment]
Rebased on master and used the batch icon. I really don't know how it should be properly resized so I just threw it into /res/drawable/

**Comment by agrajaghh** [comment]
you need to put different sizes in drawable-mdpi, hdpi, xhdpi, xxhdpi and xxxhdpi. Just look at the normal save icon

**Comment by Omikron43** [comment]
OK, done based on [this answer](https://stackoverflow.com/a/21311302/2505347).

**Comment by agrajaghh** [comment]
action bar icons are 24dp not 48dp I think :stuck_out_tongue_winking_eye:

**Comment by mcginty** [comment]
it's all good, i'll take care of the asset sizing

**Comment by mcginty** [comment]
Right now this iterates through a cursor on the main thread from what I can see, which is to be avoided. I think we'll want to use an ProgressDialogAsyncTask to collect the Attachments we'd like to save (we tend to prefer List instead of arrays), then once they're collected open a warning saying how many attachments they're about to save, and if they accept then spawn a SaveAttachmentTask.

That way all I/O happens off the main thread and the user will be more aware of how many files will be created by hitting that button.

**Comment by Omikron43** [comment]
OK, rebased and reworked according to your comments. I pluralized some strings to not have these ugly `multipleAttachments ? ... : ...` statements.
Regarding the I/O on the main thread: we know beforehand how many images will be saved (the item count of the cursor). This number is printed in the dialog but after accepting there is a very short (probably only a few millisecs with 213 images, Nexus 5) flash of the dialog collecting the images. After the flash the normal save dialog pops up.

File: src/org/thoughtcrime/securesms/MediaOverviewActivity.java [hunk 1]
```diff
@@ -137,12 +146,63 @@ public void onModified(Recipient recipient) {
...
+      public void onClick(DialogInterface dialogInterface, int i) {
+        new ProgressDialogAsyncTask<Void, Void, List<SaveAttachmentTask.Attachment>>(c,
+                                                                                     R.string.ConversationFragment_collecting_attahments,
+                                                                                     R.string.please_wait) {
+          @Override
+          protected List<SaveAttachmentTask.Attachment> doInBackground(Void... params) {
+            Cursor cursor                                   = DatabaseFactory.getImageDatabase(c).getImagesForThread(threadId);
+            List<SaveAttachmentTask.Attachment> attachments = new ArrayList<>(cursor.getCount());
+
+            cursor.moveToFirst();
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/MediaOverviewActivity.java]: while (cursor != null && cursor.moveToNext()) {

Comment (Omikron43) [inline src/org/thoughtcrime/securesms/MediaOverviewActivity.java]: done

**Comment by moxie0** [comment]
really hard to follow this when you force push a rebased update rather than adding a commit to address comments. any chance you could separate out the commit?

**Comment by fehnomenal** [comment]
@moxie0 did a bit git-fu and I think I found the original commit that you commented on and extracted the looping commit

**Comment by moxie0** [comment]
cool now as far as i can see there's only one resource in xxxhdpi, rather than sized resources for every density

**Comment by fehnomenal** [comment]
@moxie0 i had a bit trouble with finding the right sizes so @mcginty said in https://github.com/WhisperSystems/Signal-Android/pull/3975#issuecomment-159390560 that he would take care of resizing

**Comment by agrajaghh** [comment]
@fehnomenal I added the missing files and opened a PR in your repo: https://github.com/fehnomenal/TextSecure/pull/1

**Comment by fehnomenal** [comment]
nice thx

**Comment by moxie0** [comment]
in 3.12.0
